### ISSUE TITLE
timeout of asyncdispatch.drain is not respected and behaviour doesn't match poll

### ISSUE THREAD

**Issue by D-Nice** [nim-lang/Nim#14820]
The timeout behaviour appears buggy to me or at least non-intuitive, in that, if I have something async that takes 500ms, I can `poll` 5 times with 100ms timeouts, for it to resolve, whilst `drain` resolves on the 3rd call, waiting 200ms instead of the requested 100ms. The timeout will differ further, the more async events you may have in the pipeline, because of what appears to be a regression to me.

Simply said, I expected `drain` to allow events up to the timeout to complete and move on either after the timeout exceeds, or early if all pending events are complete.

My understanding was that it was meant as a replacement to `poll`, for when you'd like to allow all events up to a certain amount of timeout to complete. The commit message reads as: `introduce asyncdispatch.drain that completes all operations that can be completed immediately;` although I read `immediately` as within the timeout. The current behaviour https://github.com/nim-lang/Nim/commit/4e6f2b5313418c5f070611adb3b29a70c621782b doesn't match this at all and creates a dangerous, varying timeout, and the previous, pre-regression behaviour https://github.com/nim-lang/Nim/commit/3073f08e4857b14a6bd1b0022fe45f461bf88171 still doesn't match this and appears to behave as a carbon copy of poll.

### Example
https://play.nim-lang.org/#ix=2qch
```nim
import asyncdispatch, times

var done = false
proc somethingAsync() {.async.} =
  yield sleepAsync 5000
  echo "async done"
  done = true
  
discard somethingAsync()
while not done:
  let s0 = now()
  drain 1000
  echo "ms: ", (now() - s0).inMilliseconds
```

### Current Output
```
ms: 2002
ms: 2002
async done
ms: 997
```

### Expected Output
replacing drain with poll, one gets:
```
ms: 1001
ms: 1001
ms: 1001
ms: 1001
async done
ms: 997
```

and as it currently stands, `waitFor sleepAsync 1000` more closely matches poll, than drain does, except that the last line is 1001 as well, since it waits for this additional sleep, which is why drain is useful over just the sleepAsync hack.

### Possible Solution

* The current regression appears to have been introduced here: https://github.com/nim-lang/Nim/commit/4e6f2b5313418c5f070611adb3b29a70c621782b I think this simply needs to be reverted as a first order of business.

v1.0.6 runs closer to expected and matches poll behaviour with the above simplified example, however, it still doesn't match what I think is expected of it, which is for it to process all possible events within the given timeout. v1.0.6 will only process the timeout for as much as is needed for the first event to complete, thereby it practically matches the behaviour of poll.

Test with this more complex example with multiple asyncs: https://play.nim-lang.org/#ix=2qcm
interchange L15 between drain and poll, they both give a similar output @ v1.0.6 (1.2.0 just borks it imo):
```
async done 400
ms: 400
ms: 1001
async done 1500
async done 1500
ms: 98
ms: 1001
async done 2700
ms: 196
ms: 1001
ms: 1001
async done 5000
ms: 297
```

while something more along the lines of 
```
async done 400
ms: 1000
async done 1500
async done 1500
ms: 1001
async done 2700
ms: 1000
ms: 1001
async done 5000
ms: 998
```
was expected

a possible solution would be to revert this latest regression, but still pass `timeout` to subsequent `runOnce` calls, and decrementing `timeout` by the amount expended between each run.

### Additional Information
Related:
feat request: https://github.com/nim-lang/Nim/issues/6523
Initial PR: https://github.com/nim-lang/Nim/commit/85ac3130aa80e784fcd84c5a38122c61f4a8860d
Believed regression PR: https://github.com/nim-lang/Nim/commit/4e6f2b5313418c5f070611adb3b29a70c621782b

```
$ nim -v
Nim Compiler Version 1.2.0
# make sure to include the git hash if not using a tagged release
```

pinging @dom96 since I think you're normally interested in async portions of the codebase

**Comment by dom96** [comment]
> pinging @dom96 since I think you're normally interested in async portions of the codebase

Yeah, i'm interested. But I'm tired of `drain`, I recall fighting against its inclusion. To this day it remains a niche feature and seemingly riddled with bugs (for something as simple as this). I would honestly just remove it but I will let those that wanted it and use it to figure out how to fix the issues you've raised.

If I were you I would strongly suggest using `poll` (and if you feel you need `drain` I'd love to hear what the use case is).

CC @Araq @rayman22201 @endragor

**Comment by rayman22201** [comment]
I agree, I think @D-Nice really wants `poll` for this use case. `drain` is not mean to be a replacement for `poll` at all, and that may not be well communicated in the docs.

tbh. I forgot a lot of the context, but that change to poll was required for the async -> threadpool stuff to work iirc.
Without that change to `drain` you end up with zombie event handles and memory leaks. 

Reverting this will just bring the leaks back. Seems better to fix the bug...

See the original PR: https://github.com/nim-lang/Nim/pull/12371

[And the original discussion here](https://github.com/nim-lang/Nim/pull/12371/commits/4e6f2b5313418c5f070611adb3b29a70c621782b#r334121009)

The whole reason for `drain` to exist is because, unlike languages with event loops built into the language, Nim has explicit starting and stopping of the event loop. You need a way to reliably wait for the event loop to completely drain before you shutdown, even in situations where you may not know how many async events are left in the queue. It's a destructor of sorts imo, and really meant to only be used to drain the event loop to allow for a clean shutdown. That's the main usecase for it imo.

@dom96 if you can think of a better way to reliably drain the async event queue (that doesn't rely on exceptions), I'm all for it...

I think this is arguably outside of the original intended usage of `drain`, but this does still seem like a bug.
Particularly, it seems like the timeout is not being respected in some cases in the loop on line [1563](https://github.com/nim-lang/Nim/commit/4e6f2b5313418c5f070611adb3b29a70c621782b#diff-e7d29941cb2ca6bf4e0f5c627239bc7bR1563)...

<s>@D-Nice what platform are you running on? I don't see that info in your original post. I ask because the behavior is very dependent on the underlying ioselector implementation, which is different depending on what OS you are on.</s>

Edit: I totally missed the playground link. The environment is obviously linux b/c it's the playground. sorry :-P

**Comment by D-Nice** [comment]
@rayman22201 platform is Linux. A documentation improvement may indeed be needed, because I could very well be misunderstanding its purpose, but noted my above understanding of it. The difference I understood between it and poll, is that poll will be run once, up to either the timeout presented or the first event to finish within the timeout, while the behaviour I expected with drain was for it to go until either the timeout or ALL events complete.

Also, thanks for the additional context provided :)

@dom96 Regarding removal, unless it can be better defined that may be the way go, because its usage and purpose at this point is not clear to me, and it seems more dangerous than not, unless someone is actively using it.

So as just a general question, is the current output, indeed the expected one, where a requested time out 1000 for a single event, actually takes 2000+ms (gets way wonkier with multiple events). It seems unreliable and dangerous to me as is, and working around poll myself would appear safer unless I've missed the big picture behind drain.

Even if the definition of drain is to allow every pending operation up to the timeout sequentially to complete (which is quite weird to me), there is still a bug imo, with the runOnce call in the conditional and while loop, causing the odd single event output; simply the conditional definitely seems to need removing to me.

Basically, I have the same qualm with it as pointed out in this comment, in its current state (not saying the previous one was what I had expected exactly either): https://github.com/nim-lang/Nim/pull/12371/commits/4e6f2b5313418c5f070611adb3b29a70c621782b (building upon this comment further, it blocking longer is basically a given if the pending operation takes longer than the provided timeout).

**Comment by D-Nice** [comment]
> The whole reason for `drain` to exist is because, unlike languages with event loops built into the language, Nim has explicit starting and stopping of the event loop. You need a way to reliably wait for the event loop to completely drain before you shutdown, even in situations where you may not know how many async events are left in the queue. It's a destructor of sorts imo, and really meant to only be used to drain the event loop to allow for a clean shutdown. That's the main usecase for it imo.

The reason for drain appears to be platform specific, I'll link the comment for the apparent reason, which doesn't match the above reasoning to completely drain: https://github.com/nim-lang/Nim/issues/6523#issuecomment-337176418 which may explain why the initial implementations of it appeared to have the same effect as poll for me, since I'm running stuff mostly on Linux.

So in conclusion, I guess that implementation did what it was supposed to for macOS and Windows, but is apparently redundant for Linux. I think @endragor shedding light on this would be most useful.

**Comment by rayman22201** [comment]
@D-Nice 

> The difference I understood between it and poll, is that poll will be run once, up to either the timeout presented or the first event to finish within the timeout, while the behaviour I expected with drain was for it to go until either the timeout or ALL events complete

That's not quite correct. If you look at the implementation of runOnce (poll is just a wrapper around runOnce), you will see that it actually completes all events that are ready at the time the OS ioselector is ready (there may be multiple events ready at that time):
https://github.com/rayman22201/Nim/blob/4e6f2b5313418c5f070611adb3b29a70c621782b/lib/pure/asyncdispatch.nim#L1306

So, it's actually, Poll will run once, until the first os level ioselector event triggers (which may trigger multiple async events) or the timeout expires, which ever happens first.

Drain (is supposed to) run until **all** os level ioselector events have triggered, or the timeout expires, which ever happens first.

> there is still a bug imo

I agree. I think there is a bug here. I'm not convinced the bug is in the conditional though. I am going to spin up my local environment and run some tests and find out more.

**Comment by D-Nice** [comment]
> Drain (is supposed to) run until **all** os level ioselector events have triggered, or the timeout expires, which ever happens first.

I believe this is what I expected, except I expected a timeout of 1000ms to be respected, and not block the code for a potentially unknown timeframe.

Please check https://play.nim-lang.org/#ix=2qcm Try it with poll on 1.2.4, and try it with drain on 1.2.4, I'm interested if that is the expected behaviour. I would expect no timeout to significantly exceed 1000ms in either case.

EDIT: @dom96 please dont kill me for discarding futures :P

**Comment by rayman22201** [comment]
> I believe this is what I expected, except I expected a timeout of 1000ms to be respected, and not block the code for a potentially unknown timeframe.

I agree.
This is what makes me think it might actually be a bug in RunOnce, since asynctimers are registered as OS level signals.

**Comment by rayman22201** [comment]
@D-Nice You were right. Drain was totally broken. I opened a PR with a (hopefully) better implementation.

**Comment by D-Nice** [comment]
Since https://github.com/nim-lang/Nim/pull/14838 is merged will close this. TY!
